### PR TITLE
[WIP] Multiple improvements, fixes and implemented missing features

### PR DESCRIPTION
**Pull request by codablock**
This is an attempt to make this backend more complete. Please see individual commits for details.

I didn't commit for a long time when I did these changes and thus had to create commits from the huge bunch of pending changes in one go. This means that I can't guarantee that commits work on its own (I know for sure that they don't actually). Only the whole PR with all changes will work properly.

I can also not guarantee any compatibility to the old version. Especially if identifiers were stored externally or in plugins, these won't work anymore due to the way these were handled before. If compatibility is an issue, a fork of this repo might be the better solution so that it's very clear that there is no compatibility.

### CODE REVIEW

File: lib/bf_card.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    # WTF!? Microsoft!?
+    if color == "red":
+        return "attention"
+    elif color == "yellow":
+        return "warning"
+    elif color == "green":
+        return "good"
+    elif color == "blue":
+        return "accent"
+    return "default"
```
Comment (vasilcovsky) [inline lib/bf_card.py:11]:
```suggestion
    return {
      'red': 'attention',
      'yellow': 'warning',
      'green': 'good',
      'blue': 'accent'
    }.get(color, 'default')
```

File: botframework.py [hunk 2]
```diff
@@ -33,98 +40,87 @@ def auth(appId, appPasswd):
...
+    def conversation(self):
+        return self._conversation
 
     @property
-    def nick(self):
-        return self._name
+    def id(self):
+        if 'id' not in self._activity:
+            return None
+        return self._activity['id']
```
Comment (vasilcovsky) [inline botframework.py:113]:
```suggestion
        return self._activity.get('id')
```

File: botframework.py [hunk 3]
```diff
@@ -33,98 +40,87 @@ def auth(appId, appPasswd):
...
 
     @property
-    def fullname(self):
-        return self._name
+    def reply_to_id(self):
+        if 'replyToId' not in self._activity:
+            return None
+        return self._activity["replyToId"]
```
Comment (vasilcovsky) [inline botframework.py:119]:
```suggestion
        return self._activity.get('replyToId')
```

File: botframework.py [hunk 4]
```diff
@@ -33,98 +40,87 @@ def auth(appId, appPasswd):
...
-
-class Identifier(Person):
-    def __init__(self, obj_or_json):
-        if isinstance(obj_or_json, str):
-            subject = json.loads(obj_or_json)
-        else:
-            subject = obj_or_json
+    @property
+    def aad_object_id(self):
+        return self._conversation["aadObjectId"]
```
Comment (vasilcovsky) [inline botframework.py:80]:
```suggestion
        return self._conversation['aadObjectId']
```

File: botframework.py [hunk 5]
```diff
@@ -138,11 +134,54 @@ def __init__(self, config):
...
+        """
+        BotFramework.__build_identifier = self.build_identifier
+        for cls in (BFPerson, BFRoomOccupant, BFRoom):
+            copyreg.pickle(cls, BotFramework._pickle_identifier, BotFramework._unpickle_identifier)
+            jsonpickle.handlers.register(cls, JsonPickleHandler(self))
+
+    def _load_persistent_stuff(self):
+        self._service_url = self.get("service_url", self._default_service_url)
+
+        self.bot_account = self.get("bot_account", None)
```
Comment (vasilcovsky) [inline botframework.py:176]:
```suggestion
        self.bot_account = self.get('bot_account')
```

File: botframework.py [hunk 6]
```diff
@@ -138,11 +134,54 @@ def __init__(self, config):
...
+        if self.bot_account is not None:
+            self.bot_identifier = BFPerson.from_bf_account(self.bot_account)
+        else:
+            self.bot_identifier = None
```
Comment (vasilcovsky) [inline botframework.py:180]:
```suggestion
        self.bot_identifier = BFPerson.from_bf_account(self.bot_account) if self.bot_account else None
```

File: botframework.py [hunk 7]
```diff
@@ -151,36 +190,141 @@ def _ensure_token(self):
...
+                if "x5c" in k and "x5t" in k:
+                    b = base64.b64decode(k["x5c"][0])
+                    x5c = load_der_x509_certificate(b)
+                    self._keys[k["x5t"]] = x5c
+            self._keys_time = time.time()
+
+    def _get_key(self, x5t):
+        self._ensure_keys()
+        with self._cache_lock:
+            return self._keys.get(x5t, None)
```
Comment (vasilcovsky) [inline botframework.py:209]:
```suggestion
            return self._keys.get(x5t)
```

File: botframework.py [hunk 8]
```diff
@@ -151,36 +190,141 @@ def _ensure_token(self):
             self._token = auth(self._appId, self._appPassword)
         return self._token.access_token
 
-    def _build_reply(self, msg):
-        conversation = msg.extras['conversation']
+    def _ensure_keys(self):
+        with self._cache_lock:
+            if self._keys is not None and time.time() - self._keys_time < 60 * 60:
```
Comment (vasilcovsky) [inline botframework.py:195]:
```suggestion
            if self._keys and time.time() - self._keys_time < 60 * 60:
```

File: botframework.py [hunk 9]
```diff
@@ -151,36 +190,141 @@ def _ensure_token(self):
...
+    def _validate_jwt_token(self, auth_header):
+        try:
+            if not auth_header.startswith("Bearer "):
+                return False
+            token = auth_header[len("Bearer "):]
+            header = jwt.get_unverified_header(token)
+            if header["alg"] != "RS256" or header["typ"] != "JWT":
+                return False
+            k = self._get_key(header["x5t"])
+            if k is None:
```
Comment (vasilcovsky) [inline botframework.py:220]:
```suggestion
            if not k:
```

File: botframework.py [hunk 10]
```diff
@@ -151,36 +190,141 @@ def _ensure_token(self):
...
-            'replyToId': conversation.conversation_id,
-            'text': msg.body
+            'type': message_type,
+            'conversation': conversation.raw,
+            'service_url': self._service_url,
         }
-        return activity(conversation.reply_url, payload)
+
+        if to is not None:
+            if isinstance(to, BFPerson) or isinstance(to, BFRoomOccupant):
```
Comment (vasilcovsky) [inline botframework.py:240]:
```suggestion
            if isinstance(to, (BFPerson, BFRoomOccupant)):
```

File: botframework.py [hunk 11]
```diff
@@ -151,36 +190,141 @@ def _ensure_token(self):
...
-            'recipient': msg.frm.subject,
-            'replyToId': conversation.conversation_id,
-            'text': msg.body
+            'type': message_type,
+            'conversation': conversation.raw,
+            'service_url': self._service_url,
         }
-        return activity(conversation.reply_url, payload)
+
+        if to is not None:
```
Comment (vasilcovsky) [inline botframework.py:239]:
```suggestion
        if to:
```

File: botframework.py [hunk 12]
```diff
@@ -151,36 +190,141 @@ def _ensure_token(self):
...
+            if isinstance(to, BFPerson) or isinstance(to, BFRoomOccupant):
+                payload['recipient'] = to.to_bf_subject()
+            elif to is BFRoom:
+                pass
```
Comment (vasilcovsky) [inline botframework.py:243]:
```suggestion

```

File: botframework.py [hunk 13]
```diff
@@ -151,36 +190,141 @@ def _ensure_token(self):
...
+
+        if frm is not None:
```
Comment (vasilcovsky) [inline botframework.py:245]:
```suggestion
        if frm:
```

File: botframework.py [hunk 14]
```diff
@@ -151,36 +190,141 @@ def _ensure_token(self):
...
+            payload['from'] = frm.to_bf_subject()
+        if body is not None:
```
Comment (vasilcovsky) [inline botframework.py:247]:
```suggestion
        if body:
```

File: botframework.py [hunk 15]
```diff
@@ -151,36 +190,141 @@ def _ensure_token(self):
...
+            payload['text'] = body.lstrip()
+            payload['textFormat'] = 'markdown'
+
+        if parent_msg is not None:
```
Comment (vasilcovsky) [inline botframework.py:251]:
```suggestion
        if parent_msg:
```

File: botframework.py [hunk 16]
```diff
@@ -151,36 +190,141 @@ def _ensure_token(self):
...
+            payload["replyToId"] = conversation.id
+
+        return Activity(payload)
+
+    def _build_conversation_for_id(self, id, parent_msg):
+        if id is None:
```
Comment (vasilcovsky) [inline botframework.py:257]:
```suggestion
        if not id:
```

File: botframework.py [hunk 17]
```diff
@@ -151,36 +190,141 @@ def _ensure_token(self):
...
+        elif isinstance(id, BFRoom):
+            id2 = None
+            if parent_msg:
+                if "conversation" in parent_msg.extras and parent_msg.extras["conversation"].message_id is not None:
+                    id2 = parent_msg.extras["conversation"].id
+                elif "message_id" in parent_msg.extras:
+                    id2 = "%s;messageid=%s" % (id.room_id, parent_msg.extras["message_id"])
+                elif "conversation" in parent_msg.extras:
+                    id2 = parent_msg.extras["conversation"].id
+            if id2 is None:
```
Comment (vasilcovsky) [inline botframework.py:270]:
```suggestion
            if not id2:
```

File: botframework.py [hunk 18]
```diff
@@ -151,36 +190,141 @@ def _ensure_token(self):
...
+            return self._build_conversation_for_id(id.room, parent_msg)
+        elif isinstance(id, BFRoom):
+            id2 = None
+            if parent_msg:
+                if "conversation" in parent_msg.extras and parent_msg.extras["conversation"].message_id is not None:
+                    id2 = parent_msg.extras["conversation"].id
+                elif "message_id" in parent_msg.extras:
+                    id2 = "%s;messageid=%s" % (id.room_id, parent_msg.extras["message_id"])
+                elif "conversation" in parent_msg.extras:
+                    id2 = parent_msg.extras["conversation"].id
```
Comment (vasilcovsky) [inline botframework.py:269]: is there a way to eliminate nested conditions to reduce complexity of if statements?

File: botframework.py [hunk 19]
```diff
@@ -151,36 +190,141 @@ def _ensure_token(self):
...
+
+            return Conversation({
+                "conversationType": "channel",
+                "id": id2,
+                "isGroup": True,
+                "tenantId": id.tenant_id,
+            })
+        elif isinstance(id, BFPerson):
+            c = self.get_personal_conversation(id.person)
+            if c is None:
```
Comment (vasilcovsky) [inline botframework.py:281]:
```suggestion
            if not c:
```

**Review by vasilcovsky — COMMENTED** [review]
Thank you for your contribution!

Unfortunately, I can't merge this PR as it is because it's too big. It contains a lot of changes that I don't even know how to test. Can you split it into few pull requests?

**Comment by codablock** [comment]
@vasilcovsky Did you try to go through the PR on commit level? I understand that looking at everything at once is impossible to properly review. I tried to cut commits in a way that they are at least separated by intention.

The problem is, that I started creating commits too late, which makes it impossible for me to create intermediate PRs that would work on their own.

I will go through the comments you made so far when I find time (I hope this year...). While working on my bot, I also find more stuff that needs to be modified in the backend, for example I realised that account ids are not persistent across bot registrations (everyone gets a new id when the bot is re-registered in Azure). This made me switch to using userPrincipalName for the id, which in turn required bookkeeping of urn->account mappings.

All this looks like this PR won't be ready for quite some time. I'll keep it open and change the title to include "[WIP]". I might even decide to completely ditch everything and go for a new backend that is based on the python botframework libs...depending on how much I encounter that needs to be implemented while I see that it's a lot easier in the botframework lib.

**Comment by codablock** [comment]
I decided a very long time ago that I'm not going to use errbot and/or botframework anymore, which means I'm unable to further support this PR. I'm closing it for this reason.

**Pull request closed without merging**